File: frontend/src/api/dataUpdate.ts
/**
 * 数据更新管理API（仅管理员）
 */

import apiClient from './client'

export interface DataUpdateConfig {
  // 日K线数据更新
  auto_update_enabled: boolean
  daily_update_hour: number
  daily_update_minute: number

  // 股票列表更新
  stock_list_update_enabled: boolean
  stock_list_update_frequency: 'weekly' | 'daily'
  stock_list_update_hour: number
  stock_list_update_minute: number

  // 爬虫策略
  update_speed_preset: 'slow' | 'normal' | 'fast'
  batch_size: number
  batch_rest_seconds: number
  pre_update_random_wait: number
  incremental_update: boolean

  // 数据源配置
  data_source: 'akshare' | 'tushare'
}

export interface SchedulerStatus {
  running: boolean
  jobs: Array<{
    id: string
    name: string
    next_run_time: string | null
  }>
}

export interface ManualUpdateRequest {
  update_type: 'stock_list' | 'daily_data' | 'all'
  market?: string
  stock_codes?: string[]
}

export interface DataSourceConfig {
  data_source: 'akshare' | 'tushare'
  tushare_token: string
}

export interface DataSourceTestResult {
  success: boolean
  message: string
}

export const dataUpdateAPI = {
  // 获取配置
  getConfig: async (): Promise<DataUpdateConfig> => {
    const response = await apiClient.get<DataUpdateConfig>('/api/admin/data-update/config')
    return response.data
  },

  // 更新配置
  updateConfig: async (config: Partial<DataUpdateConfig>): Promise<DataUpdateConfig> => {
    const response = await apiClient.put<DataUpdateConfig>('/api/admin/data-update/config', config)
    return response.data
  },

  // 获取调度器状态
  getSchedulerStatus: async (): Promise<SchedulerStatus> => {
    const response = await apiClient.get<SchedulerStatus>('/api/admin/data-update/scheduler/status')
    return response.data
  },

  // 手动触发更新
  manualUpdate: async (request: ManualUpdateRequest): Promise<any> => {
    const response = await apiClient.post('/api/admin/data-update/manual-update', request)
    return response.data
  },

  // 获取数据源配置
  getDataSourceConfig: async (): Promise<DataSourceConfig> => {
    const response = await apiClient.get<DataSourceConfig>('/api/admin/data-update/data-source')
    return response.data
  },

  // 更新数据源配置
  updateDataSourceConfig: async (config: Partial<DataSourceConfig>): Promise<DataSourceConfig> => {
    const response = await apiClient.put<DataSourceConfig>('/api/admin/data-update/data-source', config)
    return response.data
  },

  // 测试数据源连接
  testDataSource: async (data_source: string): Promise<DataSourceTestResult> => {
    const response = await apiClient.post<DataSourceTestResult>(`/api/admin/data-update/data-source/test?data_source=${data_source}`)
    return response.data
  },

  // 清空指定数据源的历史数据
  clearDataBySource: async (data_source: string): Promise<{ message: string; data_source: string; deleted_count: number }> => {
    const response = await apiClient.post(`/api/admin/data-update/data-source/clear?data_source=${data_source}`)
    return response.data
  },
}

// 速度档位预设（与后端保持一致）
export const SPEED_PRESETS = {
  slow: { base: 5.0, jitterMin: 2.0, jitterMax: 5.0, label: '慢速（N100推荐）', desc: '约 7~10 秒/只' },
  normal: { base: 2.0, jitterMin: 1.0, jitterMax: 2.0, label: '标准', desc: '约 3~4 秒/只' },
  fast: { base: 0.5, jitterMin: 0.2, jitterMax: 0.5, label: '快速', desc: '约 0.7~1 秒/只' },
}

// 预估全量更新耗时（5512只股票）
export function estimateDuration(
  preset: 'slow' | 'normal' | 'fast',
  batchSize: number,
  batchRest: number,
): string {
  const totalStocks = 5512
  const speed = SPEED_PRESETS[preset]
  const avgInterval = speed.base + (speed.jitterMin + speed.jitterMax) / 2
  const totalInterval = totalStocks * avgInterval
  const batchCount = Math.floor(totalStocks / batchSize)
  const totalRest = batchCount * batchRest
  const totalSeconds = totalInterval + totalRest
  const hours = totalSeconds / 3600

  if (hours >= 10) {
    return `约 ${hours.toFixed(1)} 小时（建议夜间运行）`
  } else if (hours >= 1) {
    return `约 ${hours.toFixed(1)} 小时`
  } else {
    return `约 ${(totalSeconds / 60).toFixed(0)} 分钟`
  }
}

File: backend/app/models/data_update.py
"""数据更新配置模型"""

from pydantic import BaseModel, Field
from typing import Optional


class DataUpdateConfig(BaseModel):
    """数据更新配置"""
    # 日K线数据更新
    auto_update_enabled: bool = False
    daily_update_hour: int = 19
    daily_update_minute: int = 0

    # 股票列表更新
    stock_list_update_enabled: bool = False
    stock_list_update_frequency: str = Field(default="weekly", pattern="^(weekly|daily)$")
    stock_list_update_hour: int = 6
    stock_list_update_minute: int = 0

    # 爬虫速度档位: slow=慢速(N100), normal=标准, fast=快速
    update_speed_preset: str = Field(default="slow", pattern="^(slow|normal|fast)$")

    # 批次休息设置（N100防过热）
    batch_size: int = Field(default=50, ge=10, le=500)
    batch_rest_seconds: int = Field(default=60, ge=0, le=300)

    # 股票列表更新前随机等待（秒）
    pre_update_random_wait: int = Field(default=20, ge=0, le=120)

    # 增量更新（跳过已有最新数据的股票）
    incremental_update: bool = True

    # 数据源配置
    data_source: str = Field(default="akshare", pattern="^(akshare|tushare)$")


class DataUpdateConfigUpdate(BaseModel):
    """数据更新配置更新请求（所有字段可选）"""
    auto_update_enabled: Optional[bool] = None
    daily_update_hour: Optional[int] = None
    daily_update_minute: Optional[int] = None
    stock_list_update_enabled: Optional[bool] = None
    stock_list_update_frequency: Optional[str] = None
    stock_list_update_hour: Optional[int] = None
    stock_list_update_minute: Optional[int] = None
    update_speed_preset: Optional[str] = None
    batch_size: Optional[int] = None
    batch_rest_seconds: Optional[int] = None
    pre_update_random_wait: Optional[int] = None
    incremental_update: Optional[bool] = None
    data_source: Optional[str] = None


class ManualUpdateRequest(BaseModel):
    """手动更新请求"""
    update_type: str  # 'stock_list' | 'daily_data' | 'all'
    market: Optional[str] = None  # 仅用于 daily_data
    stock_codes: Optional[list[str]] = None  # 仅用于 daily_data


class UpdateTaskStatus(BaseModel):
    """更新任务状态"""
    task_id: str
    status: str  # 'pending' | 'running' | 'completed' | 'failed'
    message: str
    progress: Optional[int] = None  # 0-100
    total: Optional[int] = None
    completed: Optional[int] = None


class DataSourceConfig(BaseModel):
    """数据源配置"""
    data_source: str = Field(default="akshare", pattern="^(akshare|tushare)$")
    tushare_token: str = ""


class DataSourceConfigUpdate(BaseModel):
    """数据源配置更新请求"""
    data_source: Optional[str] = Field(default=None, pattern="^(akshare|tushare)$")
    tushare_token: Optional[str] = None


class DataSourceTestResult(BaseModel):
    """数据源连接测试结果"""
    success: bool
    message: str
